# Variables

- namespace
- variable types
- aliases
- whitespace

### Reminders

- In programming `=` means assignment
- There can be more than one assignment in a single line
- Anything to the right of the `=` is evaluated before assignment
    - This process proceeds from right to left

### Clicker #1

To warm our brains up and get us thinking about variable assignment again, **What would be the value of `b` after running the following code cell?**

In [2]:
a = 16
b = 'string'
c = a + 16
b = 72
my_variable = b + a
my_variable
print(b)

72


- A) 'string'
- B) 88
- C) 72
- D) This code will fail

### Clicker #2

After executing the following code, what will be the value of `var_2`?

In [3]:
var_2 = var_1 = 1

print(var_2)

1


- a) 'var_1'
- b) 1
- c) 2
- d) This code will fail

## Declaring Variables Cheat Sheet

- Names are always on the left of the `=`, values are always on the right
- Names are case sensitive
- Variables must start with letters
    - After that, they can include numbers, and underscores
    - They cannot include special characters (like &, *, #, etc)
- Python doesn't care what you name your variables
    - Humans do care. Pick names that describe the data / value that they store

## Reserved Words

There are 33 words that are not allowed to be used for variable assignment in Python 3.6.

Reserved words: False, None, True, and, as, assert, break, class, continue, def, del, elif, else, except, finally, for, from, global, if, import, in, is, lambda, nonlocal, not, or, pass, raise, return, try, while, with, yield

<center><img src="img/reserved_words.png" width="650px"></center>

In [4]:
# you will get an error if you try to assign a variable to one of these words
try = 6

SyntaxError: invalid syntax (<ipython-input-4-1c44ba76d8f9>, line 2)

## Kernels

<div class="alert alert-success">
The 'kernel' is the thing that executes your code. It is what connects the notebook (as you see it) with the part of your computer that runs code. 
</div>

Your kernel also stores your **namespace** - all the variables and code that you have declared (executed). 

It can be useful to clear and re-launch the kernel. You can do this from the 'kernel' drop down menu, at the top, optionally also clearing all ouputs. Note that this will erase any variables that are stored in memory. 

## Namespace

<div class="alert alert-success">
The namespace is the 'place' where all your currently defined code is declared - all the things you have stored in active memory. 
</div>

In [5]:
whos?

In [6]:
# You can list everything declared in the namespace with '%whos'
%whos

Variable      Type    Data/Info
-------------------------------
a             int     16
b             int     72
c             int     32
my_variable   int     88
var_1         int     1
var_2         int     1


## Variable Types

<div class="alert alert-success">
Every variable has a 'type', which refers to the kind of variable that it is, and how the computer stores that data.
</div>

In [7]:
# Declare a variable
variable_name = 1

# You can always ask Python 'what type is this variable' using:
type(variable_name)

int

### Int

<div class="alert alert-success">
`Integers` store whole numbers.
</div>

In [8]:
my_integer = 1
another_integer = 321

In [9]:
# integers can be signed
yet_another_integer = -4
type(yet_another_integer)

int

### Float

<div class="alert alert-success">
`Floats` store signed, decimal-point numbers.
</div>

In [10]:
my_float = 1.0
another_float = -231.45

In [11]:
type(another_float)

float

### String

<div class="alert alert-success">
`Strings` store characters, as text. 
</div>

In [12]:
my_string = 'words, words, words'
another_string = 'more words'

# Note that strings can be defined with either '' or ""
and_another = "and some more"

In [16]:
print(and_another)
type(and_another)

and some more


str

#### Quotation Marks

About those quotation marks...

In [14]:
my_string = 'This is a single-quoted string.'
my_string

'This is a single-quoted string.'

In [15]:
my_string = "This is a double-quoted string."
my_string

'This is a double-quoted string.'

Note that Python will put single quotes around it, even if you specify double quotes. 

A general principle is to pick something and be consistent. In this course, I'll do my best to only use single quotes.

## Boolean

<div class="alert alert-success">
`Booleans` store `True` or `False`. 
</div>

In [18]:
my_bool = True
another_bool = False

In [19]:
type(another_bool)

bool

## None

<div class="alert alert-success">
`None` is a special type that stores `None`, used to denote a null or empty value.
</div>

In [20]:
the_concept_of_nothing = None

In [21]:
type(the_concept_of_nothing)

NoneType

### Clicker #3

After executing the following code, what will the type of `var_a` be?

In [ ]:
var_a = -17.5

- A) String
- B) Int
- C) Float
- D) Boolean
- E) None

### Clicker #4

After executing the following code, what will the type of `var_b` be?

In [ ]:
var_b = '-17.5'

- A) String
- B) Int
- C) Float
- D) Boolean
- E) None

### Clicker #5

After executing the following code, what will the type of the variable `m` be?

In [23]:
n = 1
a = 'm'
m = n
type(m)

int

- A) String
- B) Int
- C) Float
- D) Boolean
- E) None

## Aliases

<div class="alert alert-success">
Variables are names assigned to a value. Values can have more than one name. 
</div>

In [24]:
# Make a variable, and an alias
a = 1
b = a
print(b)

1


Here, the value 1 is assigned to the variable `a`. We then make an alias of `a` and store that in the variable `b`. Now, the same value (1) is stored in both `a` (the original) and `b` (the alias).

What if we change the value of the original variable (`a`) - what happens to `b`?

### Clicker #6

After executing the following code, what will the values stored in `a` and `b` be?

In [25]:
# Make a variable & an alias
# change value of `a`
a = 1
b = a
a = 2

print(a)
print(b)

2
1


- A) `a` and `b` both store 1
- B) `a` and `b` both store 2
- C) `a` stores 2 `b` stores 1
- D) `a` stores 1 `b` stores 2
- E) No clue

### Reminders

- Multiple variables can relate to the same value(s)

### Mutable vs Immutable

The variable types we've talked about today are all **immutable**. This means they cannot be altered after they're created. 

In [26]:
# cannot change part of the string after creation
immutable_string = 'COGS18 is the best'
immutable_string[4] = '0'

TypeError: 'str' object does not support item assignment

Python does have **mutable** types. We'll talk about these later in the course, and these are where aliasing shines!

### Why allow aliasing? 

Aliasing can get confusing and be difficult to track, so why does Python allow it?

Well, it's more efficient to point to an alias than to make an entirely new copy of a a very large variable storing a lot of data. 

Python allows for the confusion, in favor of being more efficient.

## Indentation

Just a word on indentation after a student question last class.

Python *does* care about whitespace. 

You will get an error if it Python runs into unanticipated whitespace.

In [27]:
a = 1
    b = a
    
    print(b) 

IndentationError: unexpected indent (<ipython-input-27-b7d7efbb406d>, line 2)

There *are* times when indentation will be required and expected. We'll discuss these in future lectures.

# Operators

- assignment
- math
- comparison
- identity

## Assignment Operator

<div class="alert alert-success">
Python uses `=` for assignment.
</div>

In [28]:
my_var = 1

## Math Operators

<div class="alert alert-success">
Python uses mathematical operators `+`, `-`, `*`, `/` for 'sum', 'substract', 'multiply', and 'divide'. These operators return numbers.
</div>

In [29]:
print(2 + 3)

5


In [32]:
div_result = 4 / 2
print(div_result)
type(div_result)

2.0


float

### Order of Operations

Mathematical operators follow the rules for order of operations.

In [33]:
order_operations = 3 + 16 / 2
print(order_operations)

11.0


To specify that you want the addition to occur first, you would use parentheses.

In [34]:
specify_operations = (3 + 16) / 2
print(specify_operations)

9.5


### Clicker #7

What would be the value stored in `my_value`?

In [35]:
my_value = (3+2)+2/(16/2)
my_value

5.25

- A) 0.218
- B) 5.25
- C) 20
- D) 40
- E) Produces an error

### More Math

<div class="alert alert-success">
Python also has `**` for exponentiation and `%` for remainder (called modulus). These also return numbers.
</div>

In [36]:
# 2 to the power 3
2 ** 3

8

In [40]:
# remainder of 17 divided by 7
17 % 7

3

### Clicker #8

What would be the value stored in `modulo_time`?

In [41]:
modulo_time = 4 ** 2 % 5
modulo_time

1

- A) 0.2
- B) 1
- C) 3
- D) 3.2
- E) Produces an error